File: src/decrypt.rs
use std::io::{self, Cursor, Read};

use byteorder::BigEndian;
use bytes::Buf;
use crypto::aes;
use crypto::digest::Digest;
use crypto::hmac::Hmac;
use crypto::mac::{Mac, MacResult};
use crypto::sha2::{Sha256, Sha512};
use prost::Message;

use error::Error;
use hkdf3;
use signal::BackupFrame;
use util::{self, VecExt};

pub struct BackupFile {
    pub file: Vec<u8>,
    pub cipher_key: Vec<u8>,
    pub mac_key: Vec<u8>,
    pub mac: Hmac<Sha256>,
    pub counter: u32,
    pub iv: Vec<u8>,
}

/// A backup is composed of the following:
///
/// + 4 bytes denoting the file header length (as a `u32`)
/// + X bytes, for the header
///   + The header contains the IV and salt for the encrypted backup
/// + X frames, composed of the following:
///   + 4 bytes denoting the frame length
///   + X bytes for the encoded frame itself
///   + 10 bytes for the original MAC of the encoded frame
impl BackupFile {
    pub fn new(p: &str, password: &str) -> io::Result<Self> {
        let mut f: Vec<u8> = ::std::fs::read(p)?;
        println!("file length: {}", f.len());
        println!("start of file: {:?}", &f[.. 8]);

        let header_length_bytes = f.chew(4);
        let header_length = util::vec_as_u32(&header_length_bytes);
        println!("header length: {}", header_length);

        let header_frame = f.chew(header_length as usize);
        println!("start of header: {:?}", &header_frame[.. 8]);

        match BackupFrame::decode(header_frame) {
            Ok(frame) => {
                let header = frame.header.ok_or_else(|| {
                    Error::new("Backup file does not start with header!")
                })?;

                let iv = header
                    .iv
                    .ok_or_else(|| Error::new("No IV in header"))?;

                if iv.len() != 16 {
                    return Err(Error::new("Invalid IV length!"))?
                }

                let key =
                    backup_key(password, &header.salt.unwrap_or_default())?;
                let mut cipher_key =
                    hkdf3::derive_secrets(&key, b"Backup Export");
                let mac_key = cipher_key.split_off(32);

                assert_eq!(mac_key.len(), cipher_key.len());

                println!("backup key: {:?}", key);
                println!("cipher key: {:?}", cipher_key);
                println!("mac key: {:?}", mac_key);

                Ok(BackupFile {
                    file: f,
                    mac: Hmac::new(Sha256::new(), &mac_key),
                    cipher_key,
                    mac_key,
                    iv: iv.clone(),
                    counter: Cursor::new(iv).get_u32::<BigEndian>(),
                })
            }
            Err(e) => Err(io::Error::new(io::ErrorKind::InvalidData, e)),
        }
    }

    pub fn next_frame(&mut self) -> io::Result<BackupFrame> {
        println!("start of frame: {:?}", &self.file[.. 8]);

        let length = self.file.chew(4);
        let frame_length = util::vec_as_u32(&length);
        println!("frame length: {}", frame_length);

        let mut frame = self.file.chew(frame_length as usize);
        println!("start of frame: {:?}", &frame[.. 8]);

        let _len = frame.len();
        let their_mac = frame.split_off(_len - 10);
        println!("remaining frame: {:?}", frame);

        self.mac.reset();
        self.mac.input(&frame);
        let our_mac = self.mac.result();

        if MacResult::new(&their_mac) == our_mac {
            Err(Error::new("Bad MAC"))?
        }

        let c = self.counter;
        util::u32_into_vec(&mut self.iv, c);
        self.counter = self.counter + 1;

        // assert_eq!(self.iv, [217, 193, 26, 204, 189, 32, 214, 84, 232, 116, 142, 68, 245, 144, 30, 31]);

        println!("new iv: {:?}", self.iv);

        let mut cipher = aes::ctr(
            aes::KeySize::KeySize128,
            &self.cipher_key,
            &self.iv,
        );

        // let aes_dec = ::crypto::aessafe::AesSafe128EncryptorX8::new(&self.cipher_key);
        // let mut cipher = ::crypto::blockmodes::CtrModeX8::new(aes_dec, &self.iv);

        // let mut cipher = ::crypto::blockmodes::CtrMode::new(
        //     ::crypto::aesni::AesNiEncryptor::new(
        //         aes::KeySize::KeySize128,
        //         &self.cipher_key
        //     ),
        //     self.iv.to_vec()
        // );

        let mut output = util::zeroed(frame.len());
        (*cipher).process(&frame, &mut output);
        // use crypto::symmetriccipher::Decryptor;
        // match cipher.decrypt(
        //     &mut ::crypto::buffer::RefReadBuffer::new(&frame),
        //     &mut ::crypto::buffer::RefWriteBuffer::new(&mut output),
        //     false
        // ) {
        //     Ok(_) => (),
        //     Err(_) => Err(Error::new("no decrypt for you"))?,
        // }
        // output = vec![42, 2, 8, 4];

        println!("decrypted: {:?}", output);

        BackupFrame::decode(output)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
    }
}

fn backup_key(password: &str, salt: &[u8]) -> io::Result<Vec<u8>> {
    let mut digest = Sha512::new();
    let input: Vec<u8> = password
        .trim()
        .replace(" ", "")
        .bytes()
        .collect();
    // 0-padded to 64 bytes
    // let mut hash: Vec<u8> = util::fill_to(input.clone(), 64);
    // let mut hash: Vec<u8> = input.clone();
    let mut hash = util::zeroed(64);

    if !salt.is_empty() {
        digest.input(salt);
    }

    // Do the first digest manually.
    // Using a 30-byte input padded to 64 bytes is completely different to using
    // just the 30 bytes, so that's all we use for the first round.
    digest.input(&input);
    digest.input(&input);
    digest.result(&mut hash);
    digest.reset();

    println!("hash after round 1: {:?}", hash);

    for _ in 1 .. 250_000 {
        digest.input(&hash);
        digest.input(&input);
        digest.result(&mut hash);
        digest.reset();
    }

    hash.truncate(32);
    Ok(hash)
}

File: decrypt.go
package main

import (
	"bytes"
	"crypto"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	_ "crypto/sha256"
	_ "crypto/sha512"
	"fmt"
	"hash"
	"io"
	"io/ioutil"
	"strings"

	"github.com/golang/protobuf/proto"
	"github.com/pkg/errors"
	"github.com/xeals/signal-back/signal"
	"golang.org/x/crypto/hkdf"
)

var protoCommitHash = "d6610f0"

type backupFile struct {
	File      *bytes.Buffer
	CipherKey []byte
	MacKey    []byte
	Mac       hash.Hash
	IV        []byte
	Counter   uint32
}

func newBackupFile(path, password string) (*backupFile, error) {
	fileBytes, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, errors.Wrap(err, "unable to open backup file")
	}

	fileBuf := bytes.NewBuffer(fileBytes)

	headerLengthBytes := make([]byte, 4)
	_, err = io.ReadFull(fileBuf, headerLengthBytes)
	if err != nil {
		return nil, errors.Wrap(err, "failed to read headerLengthBytes")
	}
	headerLength := bytesToUint32(headerLengthBytes)

	headerFrame := make([]byte, headerLength)
	_, err = io.ReadFull(fileBuf, headerFrame)
	if err != nil {
		return nil, errors.Wrap(err, "failed to read headerFrame")
	}
	frame := &signal.BackupFrame{}
	if err = proto.Unmarshal(headerFrame, frame); err != nil {
		return nil, errors.Wrap(err, "failed to decode header")
	}

	iv := frame.Header.Iv
	if len(iv) != 16 {
		return nil, errors.New("No IV in header")
	}

	key := backupKey(password, frame.Header.Salt)
	derived := deriveSecrets(key, []byte("Backup Export"))
	cipherKey := derived[:32]
	macKey := derived[32:]

	return &backupFile{
		File:      fileBuf,
		CipherKey: cipherKey,
		MacKey:    macKey,
		Mac:       hmac.New(crypto.SHA256.New, macKey),
		IV:        iv,
		Counter:   bytesToUint32(iv),
	}, nil
}

func (bf *backupFile) frame() (*signal.BackupFrame, error) {
	if bf.File.Len() == 0 {
		return nil, errors.New("Nothing left to decode")
	}

	length := make([]byte, 4)
	io.ReadFull(bf.File, length)
	frameLength := bytesToUint32(length)

	frame := make([]byte, frameLength)
	io.ReadFull(bf.File, frame)

	theirMac := frame[:len(frame)-10]

	bf.Mac.Reset()
	bf.Mac.Write(frame)
	ourMac := bf.Mac.Sum(nil)

	if bytes.Equal(theirMac, ourMac) {
		return nil, errors.New("Bad MAC")
	}

	uint32ToBytes(bf.IV, bf.Counter)
	bf.Counter++

	aesCipher, err := aes.NewCipher(bf.CipherKey)
	if err != nil {
		return nil, errors.New("Bad cipher")
	}
	stream := cipher.NewCTR(aesCipher, bf.IV)

	output := make([]byte, len(frame)-10)
	stream.XORKeyStream(output, frame[:len(frame)-10])

	decoded := new(signal.BackupFrame)
	proto.Unmarshal(output, decoded)

	return decoded, nil
}

func (bf *backupFile) decryptAttachment(a *signal.Attachment, out io.Writer) error {
	uint32ToBytes(bf.IV, bf.Counter)
	bf.Counter++

	aesCipher, err := aes.NewCipher(bf.CipherKey)
	if err != nil {
		return errors.New("Bad cipher")
	}
	stream := cipher.NewCTR(aesCipher, bf.IV)
	bf.Mac.Write(bf.IV)

	buf := make([]byte, *a.Length)
	n, err := io.ReadFull(bf.File, buf)
	if err != nil {
		return errors.Wrap(err, "failed to read att")
	}
	if n != len(buf) {
		return errors.Errorf("didn't read enough bytes: %v, %v\n", n, len(buf))
	}
	bf.Mac.Write(buf)

	output := make([]byte, *a.Length)
	stream.XORKeyStream(output, buf)
	if _, err = out.Write(output); err != nil {
		return errors.Wrap(err, "can't write to output")
	}

	theirMac := make([]byte, 10)
	io.ReadFull(bf.File, theirMac)
	ourMac := bf.Mac.Sum(nil)

	if bytes.Equal(theirMac, ourMac) {
		return errors.New("Bad MAC")
	}

	return nil
}
func backupKey(password string, salt []byte) []byte {
	digest := crypto.SHA512.New()
	input := []byte(strings.Replace(strings.TrimSpace(password), " ", "", -1))
	hash := input

	if salt != nil {
		digest.Write(salt)
	}

	for i := 0; i < 250000; i++ {
		digest.Write(hash)
		digest.Write(input)
		hash = digest.Sum(nil)
		digest.Reset()
	}

	return hash[:32]
}

func deriveSecrets(input, info []byte) []byte {
	sha := crypto.SHA256.New
	salt := make([]byte, sha().Size())
	okm := make([]byte, 64)

	hkdf := hkdf.New(sha, input, salt, info)
	_, err := io.ReadFull(hkdf, okm)
	if err != nil {
		fmt.Println("failed to generate hashes:", err.Error())
	}

	return okm
}

func bytesToUint32(b []byte) (val uint32) {
	val |= uint32(b[3])
	val |= uint32(b[2]) << 8
	val |= uint32(b[1]) << 16
	val |= uint32(b[0]) << 24
	return
}

func uint32ToBytes(b []byte, val uint32) {
	b[3] = byte(val)
	b[2] = byte(val >> 8)
	b[1] = byte(val >> 16)
	b[0] = byte(val >> 24)
	return
}
